src/http/Message.cc

```diff
@@ -267,11 +267,16 @@ Http::Message::persistent() const
          */
         static SBuf close("close", 5);
         return !httpHeaderHasConnDir(&header, close);
-    } else {
+    }
+#if USE_HTTP_VIOLATIONS
+    else {
         /* for old versions of HTTP: persistent if has "keep-alive" */
         static SBuf keepAlive("keep-alive", 10);
         return httpHeaderHasConnDir(&header, keepAlive);
     }
+#endif
+    // RFC 7230 section 6.3: MUST NOT maintain a persistent connection with an HTTP/1.0 client
```

Comment:
This RFC7230 clause only applies for forward-proxy traffic. In accel mode traffic (and when act-as-origin configured) Squid is being an origin server, and for all forms of intercepted traffic (including ssl-bump) the client is expecting the same origin behaviour to be followed - per the third bullet point of section 6.3 "connection will persist after the current response".

Also, the problems this section 6.3 clause is trying to avoid are only added when the Keep-Alive header is obeyed.
> RFC 2068 "The problem was that some existing 1.0 clients may be
>   sending Keep-Alive to a proxy server that doesn't understand
>   Connection"

Squid does not use the Keep-Alive header field, so there are no bugs introduced by the current behaviour. The same cannot be said if we start ignoring the Connection:keep-alive from HTTP/1.0 clients. Past reports both to us and to other proxy vendors indicate that these types of client very often cannot handle their followup pipelined requests being ignored when they expect an origin to exist.

So, we need the above cases to obey the Connection:keep-alive header regardless of violations being disabled.

Reply:
> Squid does not use the Keep-Alive header field, so there are no bugs introduced by the current behaviour.

I am afraid that is not correct. For example, HttpStateData::httpBuildRequestHeader() adds this header field:

```
if (flags.keepalive) {
        hdr_out->putStr(Http::HdrType::CONNECTION, "keep-alive");
    }
```

Reply:
That is adding Connection:keep-alive header. The issue is about a header whose name is "Keep-Alive".
see https://tools.ietf.org/html/draft-thomson-hybi-http-timeout which tries to codify what the header is being used for and why its dangerous.

Reply:
@yadij, from your initial comment it is not clear what to do with this improvement.
To summarize, Squid (in its proxy mode) violates an RFC7230 MUST requirement and we can either:
* Fix Squid accordingly
* State that this MUST is wrong (inapplicable to Squid) and, hence, fix the RFC.

In (1) case we would need to adjust the code(as you noted) so that the requirement was applied only in forward proxy mode (act-as-origin, accel etc.).

Reply:
Summarizing, I think that future development would require much more time spent on non-trivial changes while the improvement is hardly noticeable (since nobody actually suffers from this bug). So, if my simple solution for "USE_HTTP_VIOLATIONS" mode is not acceptable, I would "do nothing", as you suggest.

Reply:
Okay, I do not think the simple change is enough to accept. So if the more complex is not going to happen now please close this PR.